File: Stacks/Stacks_using_queues.py
''' Name : Abhinav kumar
Github username : Abhinavcode13
Repository name : data-structures-and-algorithms
Problem : Implement Stack using Queues in Python
Issue Number : #261
Problem statement :

Explanation of the below python code :

__init__(self) initializes the two queues (q1 and q2) in the constructor.

push(self, x: int) adds a new element x to the top of the stack. It first adds the element to q2, and then moves all the elements from q1 to q2 using a while loop. Finally, it swaps q1 and q2 so that q1 always contains the elements in the stack.

pop(self) -> int removes and returns the element at the top of the stack, which is the first element in q1.

top(self) -> int returns the element at the top of the stack, which is also the first element in q1.

empty(self) -> bool returns True if the stack is empty (i.e., q1 is empty), and False otherwise.

The deque data structure from the collections module is used to implement the two queues, since it provides efficient O(1) operations for adding and removing elements from both ends of the queue.

Overall, this implementation is quite simple and efficient, since the push and pop operations take O(1) time complexity on average, while the top and empty operations take constant time.

-------------------------------------------------------------------------//Python code begins here----------------------------------------------------------------------------
'''

from collections import deque

class Stack:
    def __init__(self):
        self.queue = deque()

    def push(self, x: int) -> None:
        self.queue.append(x)
        # Move all existing elements to the end of the queue
        for _ in range(len(self.queue) - 1):
            self.queue.append(self.queue.popleft())

    def pop(self) -> int:
        return self.queue.popleft()

    def top(self) -> int:
        return self.queue[0]

    def empty(self) -> bool:
        return len(self.queue) == 0

# Example usage
stack = Stack()
while True:
    print("1. Push")
    print("2. Pop")
    print("3. Top")
    print("4. Empty")
    print("5. Quit")
    choice = int(input("Enter your choice: "))
    if choice == 1:
        x = int(input("Enter the element to push: "))
        stack.push(x)
        print("Element pushed to stack")
    elif choice == 2:
        if stack.empty():
            print("Stack is empty")
        else:
            x = stack.pop()
            print("Popped element:", x)
    elif choice == 3:
        if stack.empty():
            print("Stack is empty")
        else:
            x = stack.top()
            print("Top element:", x)
    elif choice == 4:
        if stack.empty():
            print("Stack is empty")
        else:
            print("Stack is not empty")
    elif choice == 5:
        break
    else:
        print("Invalid choice")

File: Stacks/stacks_using_queues.java
/* Name : Abhinav kumar
Github username : Abhinavcode13
Repository name : data-structures-and-algorithms
Problem : Implement Stack using Queues in Java
Issue Number : #259
Problem statement :

Explanation of the below Java code :

In the above code, the main method demonstrates the usage of the stack by providing the user with a menu to select the operation they want to perform - push, pop, or quit. The user can input the value to push, and the program prints the pushed or popped value accordingly.

The program uses the Scanner class to take user input from the console. The push method adds the element to the first queue (q1) and updates the top variable. The pop method transfers all elements except the last one from q1 to q2, removes the last element from q1 and swaps the two queues. The peek method returns the top variable, and the isEmpty method checks if q1 is empty.

When the user selects the push operation, the program prompts the user to enter the value to push, and it calls the push method to push the value onto the stack. When the user selects the pop operation, the program checks if the stack is empty and prints an error message if it is. Otherwise, it calls the pop method to pop the value from the stack and prints it.

The program continues to prompt the user for input until the user selects the quit operation


*/
-------------------------------------------------------------------------------------------------------//Java code begins here--------------------------------------------------------------------------------------------------------------------------

import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class StackUsingQueues {
// Create two queues as instance variables
private Queue<Integer> q1 = new LinkedList<>();
private Queue<Integer> q2 = new LinkedList<>();

// Create a variable to hold the top element of the stack
private int top;

// Method to push an element onto the stack
public void push(int x) {
    q1.add(x); // Add the element to the first queue
    top = x; // Update the top variable to hold the new element
}

// Method to pop an element from the stack
public int pop() {
    // Move all elements except the last one from the first queue to the second queue
    while (q1.size() > 1) {
        top = q1.remove();
        q2.add(top);
    }
    // Remove the last element from the first queue, which is the element to be popped
    int popValue = q1.remove();
    // Swap the queues so that the second queue becomes the first queue for the next operation
    Queue<Integer> temp = q1;
    q1 = q2;
    q2 = temp;
    return popValue; // Return the popped element
}

// Method to peek at the top element of the stack
public int peek() {
    return top; // Return the top element of the stack
}

// Method to check if the stack is empty
public boolean isEmpty() {
    return q1.isEmpty(); // Return whether the first queue is empty
}

// Main method to run the program
public static void main(String[] args) {
    // Create a new instance of the StackUsingQueues class
    StackUsingQueues stack = new StackUsingQueues();
    // Create a new Scanner object to read input from the user
    Scanner scanner = new Scanner(System.in);
    // Create a loop to continuously prompt the user for input
    while (true) {
        System.out.println("Select operation -\n"
                + "1. Push\n"
                + "2. Pop\n"
                + "3. Quit");
        // Read the user's choice
        int choice = scanner.nextInt();
        // Check the user's choice and perform the corresponding operation
        if (choice == 1) {
            System.out.print("Enter value to push: ");
            int val = scanner.nextInt();
            stack.push(val);
            System.out.println("Pushed value: " + val);
        } else if (choice == 2) {
            if (stack.isEmpty()) {
                System.out.println("Stack is empty.");
            } else {
                int val = stack.pop();
                System.out.println("Popped value: " + val);
            }
        } else if (choice == 3) {
            break;
        } else {
            System.out.println("Invalid choice. Please try again.");
        }
    }
}
}
